docs/advanced/pycpp/object.rst

```diff
@@ -168,3 +173,77 @@ Generalized unpacking according to PEP448_ is also supported:
     Python functions from C++, including keywords arguments and unpacking.
 
 .. _PEP448: https://www.python.org/dev/peps/pep-0448/
+
+.. _pytypes_interaction_with_none:
+
+Interaction with None
+=====================
+
+You may be tempted to use types like ``py::str`` and ``py::dict`` in C++
+signatures (either pure C++, or in bound signatures). However, there are some
+"gotchas" for ``py::none()`` and how it interacts with these types. In best
+case scenarios, it will fail fast (e.g. with default arguments); in worst
+cases, it will silently work but corrupt the types you want to work with.
+
+In general, the pytypes like ``py::str``, ``py::dict``, etc., are
+strict **non-nullable** reference types. They may not store a copy when
+assigned to, but they cannot store ``None``. For statically typed languages,
+this is in contrast  Java's ``String`` or ``List<E>``, or C#'s ``string`` or
+``List<T>``, which are strict nullable refernce types, and C++'s
+``std::string``, which is simply a value type, or
+``std::optional<std::string>``, which is a nullable value type.
+
+.. note::
+
+    You actually *can* make ``py::str``, ``py::dict``, etc. nullable by
```

Comment:
Er, I dunno if this is actually true. I know that the "is default constructed" predicates are used throughout the pybind codebase for easy checks (I think), but I'm not actually sure how `type_caster`s will actually handle those (e.g. fail fast, admit weird leaky behavior, or work fine without a hitch).

Reply:
The default constructor doesn't construct a valid python object. I'm thinking that this might cause more confusion that it will help. Maybe we shouldn't mention this at all.

Reply:
Indeed, this is only on a C++ level. There's a difference between `None` and "no object".

Reply:
I think it should be mentioned somewhere... but yeah, perhaps here is not the best place. Will mull it over.

Reply:
Done. Still mentioned, but separated it into a separate section to make it less confusing.